### PR TITLE
Add detail to CUDA detections, better for issue reports

### PR DESCRIPTION
**Pull request by Spudz76**
Maybe this can make it into todays release? Simple output changes that tell us more about CUDA during runtime detection.  I've been running it on my rigs (on top of latest dev) for months, it's nice to have the details whilst also not being too verbose.

### CODE REVIEW

**Comment by Spudz76** [comment]
Would have used logger instance but it is not imported in the CUDA code and wasn't sure if it could be anyway.  Therefore regular `printf()` and no nice standard date stamp headering...

**Comment by Spudz76** [comment]
Output sample:
```
[2018-05-20 20:59:34] : Mining coin: monero7
[2018-05-20 20:59:34] : Starting NVIDIA GPU thread 0, no affinity.
CUDA [8.0/8.0] GPU#0, arch 20: "GeForce GTX 465"... init succeeded
[2018-05-20 20:59:34] : Starting NVIDIA GPU thread 1, no affinity.
CUDA [8.0/8.0] GPU#1, arch 21: "GeForce GT 520"... init succeeded
[2018-05-20 20:59:34] : Starting NVIDIA GPU thread 2, no affinity.
CUDA [8.0/8.0] GPU#2, arch 21: "GeForce GTX 550 Ti"... init succeeded
[2018-05-20 20:59:34] : Starting NVIDIA GPU thread 3, no affinity.
CUDA [8.0/8.0] GPU#3, arch 21: "GeForce GTX 550 Ti"... init succeeded
[2018-05-20 20:59:34] : Compiling code and initializing GPUs. This will take a while...
[snip]
```

**Comment by Spudz76** [comment]
The version output `[8.0/8.0]` is `[driver/api]` same as the format used by Claymore and probably other apps as well (pseudo-standard?).

**Comment by Spudz76** [comment]
Tested on Windows 7 + CUDA 9.2 + 397.64

Windows 7 driver 396.xx versions with initial CUDA 9.2 support were not very stable in general (even gaming).  Linux seems to be about as bad, and does not yet have a 397.xx version to try.

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 1]
```diff
@@ -472,7 +480,7 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 		std::vector<int>::iterator it = std::find(arch.begin(), arch.end(), 20);
 		if(it == arch.end())
 		{
-			printf("WARNING: NVIDIA GPU %d: miner not compiled for CUDA architecture %d.\n", ctx->device_id, gpuArch);
+			printf("SKIP! (arch not included)\n");
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: This removes important information, please revert this change.

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 2]
```diff
@@ -490,7 +498,7 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 				minSupportedArch = arch[i];
 		if(minSupportedArch < 30 || gpuArch < minSupportedArch)
 		{
-			printf("WARNING: NVIDIA GPU %d: miner not compiled for CUDA architecture %d.\n", ctx->device_id, gpuArch);
+			printf("SKIP! (arch not included)\n");
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: please revert this

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 3]
```diff
@@ -556,14 +564,14 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 		err = cudaSetDevice(ctx->device_id);
 		if(err != cudaSuccess)
 		{
-			printf("WARNING: NVIDIA GPU %d: cannot be selected.\n", ctx->device_id);
+			printf("SKIP! (cudaSetDevice failed)\n");
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: please revert this, device id is important

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 4]
```diff
@@ -556,14 +564,14 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
...
 			return 2;
 		}
 		// trigger that a context on the gpu will be allocated
 		err = cudaMalloc(&tmp, 256);
 		if(err != cudaSuccess)
 		{
-			printf("WARNING: NVIDIA GPU %d: context cannot be created.\n", ctx->device_id);
+			printf("SKIP! (cudaMalloc failed)\n");
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: please revert this, also the WARNING is important because the word warning is somewhere defined in the docs.

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 5]
```diff
@@ -596,9 +604,7 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 		size_t usedMem = totalMemory - freeMemory;
 		if(usedMem >= maxMemUsage)
 		{
-			printf("WARNING: NVIDIA GPU %d: already %s MiB memory in use, skip GPU.\n",
-				ctx->device_id,
-				std::to_string(usedMem/byteToMiB).c_str());
+			printf("SKIP! (already %s MiB memory in use)\n", std::to_string(usedMem/byteToMiB).c_str());
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: please revert this

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 6]
```diff
@@ -627,6 +633,7 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 		}
 
 	}
+	printf("init succeeded\n");
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: please revert this, there is not information in it because the device id and the backend is missing. Better remove this.

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu (hunk 1 above)
Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: Duplicate information once it dumps all information in the detection "header"
Note there is no newline on the detect header so it would actually be on the same line for that matter.

Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: Point of patch being, we get all that info all the time, not just when something is caught.
So dumping the info twice is pointless?

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu (hunk 2 above)
Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: Same as above, duplicate once the header is in

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu (hunk 3 above)
Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: Same as above, duplicate information, it's literally on the same line with the device_id and arch regardless if error or not (and forms a sentence...)

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu (hunk 4 above)
Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: But it literally skips the GPU then?
Warnings are for a case where it can continue but degraded (no-AES)
You can't warning while also taking the failure action (no GPU init).

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu (hunk 6 above)
Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: NOTE the newlines this is PART OF THE SAME OUTPUT LINE sir.

**Comment by Spudz76** [comment]
Propose patch that fills a string buffer and printf() at the end for exact same behavior.
Which would force the line to be a contiguous line, even though it's always a line anyways (nothing threads and nukes the output stream, at detection time).

```
[2018-05-20 20:59:34] : Starting NVIDIA GPU thread 3, no affinity.
CUDA [8.0/8.0] GPU#3, arch 21: "GeForce GTX 550 Ti"... init succeeded
```
^^^ see init succeeded, same line, all part of unified message

```
[2018-05-20 20:59:34] : Starting NVIDIA GPU thread 3, no affinity.
CUDA [8.0/8.0] GPU#3, arch 21: "GeForce GTX 550 Ti"... SKIP! (cudaMalloc failed)
```
^^^ see how it would look when choking on memory? same line, why print `#3 #3 #3 #3 #3` too many times?

**Comment by Spudz76** [comment]
I want it to be super clear it SKIPPED AND IGNORED
WARNING says trying to work anyway, but if it's not init then it's not there and usable, thus not a warning but a fatal, which is ignored thus SKIP.
I would prefer to adjust the docs if that's what's wrong now.
This is not wrong.

**Comment by Spudz76** [comment]
Sorry if I was a touch gruff and rough
But the whole special coolness of this patch is that all that garbage is rolled up into a line not barfing 3 lines with the same GPU#X which looks messy and is less readable.  So your "undo the main reason for this" directives rubbed me wrong.  Apologies.

**Comment by psychocrypt** [comment]
could you please rebase this PR, that I can have again a look to it. Sry for beeing so late.

**Comment by Spudz76** [comment]
Yeah, it's a huge mess now, I'll fix up soon.

**Comment by Spudz76** [comment]
Applied requested changes with minor improvement of duplicated strings

New output sample:
```
[2018-07-13 19:34:00] : Starting NVIDIA GPU thread 0, no affinity.
CUDA [8.0/8.0] GPU#0, device architecture 20: "GeForce GTX 465"... device init succeeded
[2018-07-13 19:34:00] : Starting NVIDIA GPU thread 1, no affinity.
CUDA [8.0/8.0] GPU#1, device architecture 21: "GeForce GTX 550 Ti"... device init succeeded
[2018-07-13 19:34:00] : Starting NVIDIA GPU thread 2, no affinity.
CUDA [8.0/8.0] GPU#2, device architecture 21: "GeForce GTX 550 Ti"... device init succeeded
```

**Comment by psychocrypt** [comment]
Shit I forget that we have this PR. I will check it tonight.

**Comment by Spudz76** [comment]
is cool, I probably could tag you more often especially since the changes will never notify as completed / automatically yay GitHub

shit I forget to tag people haha

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 7]
```diff
@@ -496,13 +504,14 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 	while ( ss >> tmpArch )
 		arch.push_back( tmpArch );
 
+	#define MSG_CUDA_NO_ARCH "WARNING: skip device - binary does not contain required device architecture\n"
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: Please do not use a define for this. Please keep the message duplicated or use a intermediate variable.

File: xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu [hunk 8]
```diff
@@ -582,18 +591,19 @@ extern "C" int cuda_get_deviceinfo(nvid_ctx* ctx)
 
 		int* tmp;
 		cudaError_t err;
+		#define MSG_CUDA_FUNC_FAIL "WARNING: skip device - %s failed\n"
```
Comment (psychocrypt) [inline xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: some here please do not use defines here.

Comment (Spudz76) [reply xmrstak/backend/nvidia/nvcc_code/cuda_extra.cu]: Weird, ok.  #define is functionally identical to duplicated static strings by the time it compiles, with bonus of never being able to screw up the message separately...  allocating string on stack is wasteful so the define was the best way out while not having dupe message strings.  Also maybe sort of thinking into the future and translation packs which would essentially require defines.
